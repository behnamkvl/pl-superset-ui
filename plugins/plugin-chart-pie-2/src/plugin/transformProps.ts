/**
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
import { ChartProps } from '@superset-ui/core';
import { PieChartData, RadarProps } from '../PieChart';
import { LegendPosition } from '../utils';

type FormData<G extends string, DK extends string> = {
  colorScheme: string;
  groupby: G;
  metric: {
    label: DK;
  };
  colorPicker: string;
  showLegend: boolean;
  labelType: string;
  showLabels: boolean;
  legendPosition: LegendPosition;
};

export default function transformProps<G extends string, DK extends string>(chartProps: ChartProps): RadarProps<G, DK> {
  const { name, stroke, fill, width, height, formData, queriesData } = chartProps;
  const {
    colorScheme,
    groupby,
    metric,
    colorPicker,
    showLegend,
    labelType,
    showLabels,
    legendPosition,
  } = formData as FormData<G, DK>;
  const data = queriesData[0].data as PieChartData<G, DK>[];

  return {
    name,
    stroke,
    fill,
    dataKey: metric.label,
    width,
    legendPosition,
    height,
    data: data.filter(item => item[metric.label] !== null).filter(item => item[groupby] !== null),
    baseColor: colorPicker,
    colorScheme,
    showLegend,
    showLabels: showLabels,
    groupBy: groupby,
    labelType,
  };
}
