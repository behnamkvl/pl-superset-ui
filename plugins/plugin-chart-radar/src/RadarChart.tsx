/**
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
import React, { useState, FC, useEffect, memo, useRef, useCallback } from 'react';
import { styled, t, CategoricalColorNamespace } from '@superset-ui/core';
import {
  RadarChart as RechartsRadarChart,
  Radar as RechartsRadar,
  PolarGrid,
  PolarAngleAxis,
  PolarRadiusAxis,
  Cell,
  RechartsFunction,
  Legend,
  RadarProps as RechartsRadarProps,
  LegendType,
} from 'recharts';
import { LegendPosition, renderActiveShape, getLegendProps, ActiveShapeProps } from './utils';

export const LABELS_MARGIN = 100;

type EventData = {
  color: string;
  id: string;
  type: LegendType;
  value: string;
};

type RadarStylesProps = {
  height: number;
  width: number;
  legendPosition: LegendPosition;
};

type GroupBy<G extends string> = Record<G, string>;

export type RadarChartData<G extends string, DK extends string> = GroupBy<G> & Record<DK, number>;

export type RadarProps<G extends string = string, DK extends string = string> = {
  name: string,
  stroke: string,
  fill: string,
  height: number;
  width: number;
  data: RadarChartData<G, DK>[];
  dataKey: DK;
  dataKey2: DK;
  onClick?: RechartsFunction;
  colorScheme: string;
  baseColor: string;
  legendPosition: LegendPosition;
  showLegend: boolean;
  showLabels: boolean;
  groupBy: G;
  labelType: string;
};

const Notification = styled.div`
  cursor: pointer;
  width: 100%;
  display: flex;
  justify-content: center;
  align-items: center;
  padding: ${({ theme }) => theme.gridUnit * 4}px;
  border-radius: ${({ theme }) => theme.gridUnit * 2}px;
  color: ${({ theme }) => theme.colors.info.dark1};
  background-color: ${({ theme }) => theme.colors.info.light1};
`;

const Styles = styled.div<RadarStylesProps>`
  border-radius: ${({ theme }) => theme.gridUnit * 2}px;
  height: ${({ height }) => height};
  width: ${({ width }) => width};
  overflow: auto;

  & .recharts-legend-item {
    cursor: pointer;
    white-space: nowrap;
  }
`;

const RadarChart: FC<RadarProps<string, string>> = memo(props => {
  const {
    name,
    stroke,
    fill,
    dataKey,
    dataKey2,
    data,
    height,
    width,
    colorScheme,
    showLegend,
    showLabels,
    groupBy,
    labelType,
    legendPosition,
  } = props;
  const [notification, setNotification] = useState<string | null>(null);
  const [disabledDataKeys, setDisabledDataKeys] = useState<string[]>([]);
  const [legendWidth, setLegendWidth] = useState<number | null>(0);
  const [updater, setUpdater] = useState<number>(0);
  const rootRef = useRef<HTMLDivElement>(null);
  const isSideLegend =
    showLegend && (legendPosition === LegendPosition.right || legendPosition === LegendPosition.left);

  const currentData = data.filter(item => !disabledDataKeys.includes(item[groupBy]));
  const isExplore = window.location.pathname.includes('/explore');

  useEffect(() => {
    if (isSideLegend && rootRef.current) {
      const legend = rootRef.current.querySelector('.recharts-legend-wrapper');
      const currentWidth = legend?.getBoundingClientRect()?.width || null;
      if (currentWidth !== legendWidth) {
        setLegendWidth(currentWidth ? currentWidth + 20 : currentWidth);
      }
    }
  }, [updater, isSideLegend, legendWidth]);

  useEffect(() => {
    if (isSideLegend) {
      setLegendWidth(0);
    }
  }, [isSideLegend, props]);

  const forceUpdate = useCallback(() => setUpdater(Math.random()), []);

  useEffect(() => {
    forceUpdate();
  }, [forceUpdate, props]);

  const closeNotification = () => setNotification(null);


  const chartMargin = showLabels ? LABELS_MARGIN : 20;

  const handleLegendClick = ({ id }: EventData) => {
    let resultKeys = [];
    if (disabledDataKeys.includes(id)) {
      resultKeys = disabledDataKeys.filter(item => item !== id);
    } else {
      resultKeys = [...disabledDataKeys];
      resultKeys.push(id);
    }
    setDisabledDataKeys(resultKeys);
  };

  const chartHeight = height;
  const outerRadius = (width < chartHeight ? width : chartHeight) / 2 - chartMargin;
  const chartWidth =
    isSideLegend && legendWidth ? Math.max((outerRadius + chartMargin) * 2 + legendWidth, width) : width;

  const radarProps: RechartsRadarProps & { key?: string | number } = {
    // name: 'aaaa',
    // stroke: '#8884d8',
    // fill: '#8884d8',
    name,
    stroke,
    fill,
    key: updater,
    data: currentData,
    dataKey,
    dataKey2,
    label: showLabels
      ? labelProps => renderActiveShape({ ...labelProps, groupBy, labelType } as ActiveShapeProps)
      : false,
  };


  return (
    <Styles height={height} width={width} legendPosition={legendPosition} ref={rootRef}>
      {notification && <Notification onClick={closeNotification}>{notification}</Notification>}
      <RechartsRadarChart key={updater} width={chartWidth} height={height} data={currentData} >
      <PolarGrid />
      <PolarAngleAxis dataKey={groupBy} />
      <PolarRadiusAxis angle={30} domain={[0, 150]} />
        {/* {showLegend && (
          <Legend
            onClick={handleLegendClick}
            {...getLegendProps(legendPosition, height, legendWidth)}
            iconType="square"
            iconSize={10}
            payload={data.map(item => ({
              value: item[groupBy],
              id: item[groupBy],
              payload: item,
              type: disabledDataKeys.includes(item[groupBy]) ? 'line' : 'circle',
              color: CategoricalColorNamespace.getScale(colorScheme)(item[groupBy]),
            }))}
          />
        )} */}
        {/* {((isSideLegend && legendWidth) || !isSideLegend) && (
          <RechartsRadar {...radarProps} stroke="#82ca9d" fill="#82ca9d" fillOpacity={0.6}>
          </RechartsRadar>
        )} */}
      <RechartsRadar dataKey={radarProps.dataKey} stroke="#1df5a9" fill="#6ddeb6" fillOpacity={0.1} />
      <RechartsRadar dataKey={radarProps.dataKey2} stroke="#f77e8c" fill="#f7a1ab" fillOpacity={0.1} />
      <Legend />
      </RechartsRadarChart>
    </Styles>
  );
});

export default RadarChart;
