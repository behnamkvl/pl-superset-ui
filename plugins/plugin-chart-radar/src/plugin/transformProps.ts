/**
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
import { ChartProps } from '@superset-ui/core';
import { RadarChartData, RadarProps } from '../RadarChart';
import { LegendPosition } from '../utils';

type MetricObject<DK extends string> = {
  label: DK;
  column: {
    verbose_name: DK;
  };
};

type FormData<G extends string, DK extends string> = {
  colorScheme: string;
  groupby: G;
  metric: {
    label: DK;
  };
  colorPicker: string;
  showLegend: boolean;
  labelType: string;
  showLabels: boolean;
  legendPosition: LegendPosition;
  customMetrics: MetricObject<DK>[];
};

export default function transformProps<G extends string, DK extends string>(chartProps: ChartProps): RadarProps<G, DK> {
  const { name, stroke, fill, width, height, formData, queriesData } = chartProps;
  const customMetrics = formData.customMetrics.map(({ label }) => label).sort();
  const {
    colorScheme,
    groupby,
    metric,
    colorPicker,
    showLegend,
    labelType,
    showLabels,
    legendPosition,
  } = formData as FormData<G, DK>;
  const data = queriesData[0].data as RadarChartData<G, DK>[];

  return {
    name,
    stroke,
    fill,
    dataKey: metric.label,
    dataKey2: customMetrics[1],
    width,
    legendPosition,
    height,
    data: data.filter(item => item[metric.label] !== null).filter(item => item[groupby] !== null),
    baseColor: colorPicker,
    colorScheme,
    showLegend,
    showLabels: showLabels,
    groupBy: groupby,
    labelType,
  };
}
