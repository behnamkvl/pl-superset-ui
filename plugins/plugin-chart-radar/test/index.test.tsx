/**
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
import React from 'react';
import { render } from '@testing-library/react';
import * as recharts from 'recharts';
import { supersetTheme, ThemeProvider } from '@superset-ui/core';
import RadarChartPlugin from '../src';
import transformProps from '../src/plugin/transformProps';
import { categoryPercentageDonutNoLegend, legendLeftWithoutLabels, legendTopPercentage } from './__mocks__/radarProps';
import RadarChart from '../src/RadarChart';

jest.mock('recharts');

describe('plugin-chart-piw', () => {
  const RechartsRadarChart = jest.fn(props => <div {...props} />);
  const Legend = jest.fn(() => <div />);
  const Radar = jest.fn(() => <div />);
  beforeEach(() => {
    // Recharts still have some UNSAFE react functions that failing test
    jest.spyOn(console, 'warn').mockImplementation(() => null);

    jest.clearAllMocks();

    // @ts-ignore
    recharts.RadarChart = RechartsRadarChart;
    // @ts-ignore
    recharts.Radar = Radar;
    // @ts-ignore
    recharts.Legend = Legend;
  });

  const getWrapper = (props: object) =>
    render(
      <ThemeProvider theme={supersetTheme}>
        {/*
       // @ts-ignore (no need emulate all props) */}
        <RadarChart {...transformProps(props)} />
      </ThemeProvider>,
    );

  it('exists', () => {
    expect(RadarChartPlugin).toBeDefined();
  });

  it('Chart with legend top / percentage labels ', () => {
    getWrapper(legendTopPercentage);
    expect({
      RadarChartProps: RechartsRadarChart.mock.calls[1],
      RadarProps: Radar.mock.calls[1],
      LegendProps: Legend.mock.calls[1],
    }).toMatchSnapshot();
  });

  it('Chart with legend left / without labels', () => {
    getWrapper(legendLeftWithoutLabels);
    expect({
      RadarChartProps: RechartsRadarChart.mock.calls[1],
      RadarProps: Radar.mock.calls[1],
      LegendProps: Legend.mock.calls[1],
    }).toMatchSnapshot();
  });

  it('Chart Donut with category and percentage labels / without legend', () => {
    getWrapper(categoryPercentageDonutNoLegend);
    expect({
      RadarChartProps: RechartsRadarChart.mock.calls[1],
      RadarProps: Radar.mock.calls[1],
    }).toMatchSnapshot();
    expect(Legend).not.toBeCalled();
  });
});
