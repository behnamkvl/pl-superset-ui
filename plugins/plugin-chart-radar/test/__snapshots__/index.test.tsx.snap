// Jest Snapshot v1, https://goo.gl/fbAQLP


exports[`plugin-chart-piw Chart with legend left / without labels 1`] = `
Object {
  "LegendProps": Array [
    Object {
      "align": "left",
      "iconSize": 10,
      "iconType": "square",
      "layout": "vertical",
      "onClick": [Function],
      "payload": Array [
        Object {
          "color": "#1f77b4",
          "id": "2020",
          "payload": Object {
            "AVG(efficiency)": 0.0566666666666667,
            "subject": "2020",
            "fullMark": 150,
          },
          "type": "circle",
          "value": "2020",
        },
        Object {
          "color": "#ff7f0e",
          "id": "2018",
          "payload": Object {
            "AVG(efficiency)": 0.038,
            "subject": "2018",
            "fullMark": 150,
          },
          "type": "circle",
          "value": "2018",
        },
        Object {
          "color": "#2ca02c",
          "id": "2019",
          "payload": Object {
            "AVG(efficiency)": 0.0162068965517241,
            "subject": "2019",
            "fullMark": 150,
          },
          "type": "circle",
          "value": "2019",
        },
        Object {
          "color": "#d62728",
          "id": "Total subject",
          "payload": Object {
            "AVG(efficiency)": 0.0138888888888889,
            "subject": "Total subject",
            "fullMark": 150,
          },
          "type": "circle",
          "value": "Total subject",
        },
        Object {
          "color": "#9467bd",
          "id": "2017",
          "payload": Object {
            "AVG(efficiency)": 0.011304347826087,
            "subject": "2017",
            "fullMark": 150,
          },
          "type": "circle",
          "value": "2017",
        },
      ],
      "verticalAlign": "middle",
      "wrapperStyle": Object {
        "maxHeight": 400,
        "width": 0,
      },
    },
    Object {},
  ],
  "RadarChartProps": Array [
    Object {
      "children": Array [
        <mockConstructor
          align="left"
          iconSize={10}
          iconType="square"
          layout="vertical"
          onClick={[Function]}
          payload={
            Array [
              Object {
                "color": "#1f77b4",
                "id": "2020",
                "payload": Object {
                  "AVG(efficiency)": 0.0566666666666667,
                  "subject": "2020",
                  "fullMark": 150,
                },
                "type": "circle",
                "value": "2020",
              },
              Object {
                "color": "#ff7f0e",
                "id": "2018",
                "payload": Object {
                  "AVG(efficiency)": 0.038,
                  "subject": "2018",
                  "fullMark": 150,
                },
                "type": "circle",
                "value": "2018",
              },
              Object {
                "color": "#2ca02c",
                "id": "2019",
                "payload": Object {
                  "AVG(efficiency)": 0.0162068965517241,
                  "subject": "2019",
                  "fullMark": 150,
                },
                "type": "circle",
                "value": "2019",
              },
              Object {
                "color": "#d62728",
                "id": "Total subject",
                "payload": Object {
                  "AVG(efficiency)": 0.0138888888888889,
                  "subject": "Total subject",
                  "fullMark": 150,
                },
                "type": "circle",
                "value": "Total subject",
              },
              Object {
                "color": "#9467bd",
                "id": "2017",
                "payload": Object {
                  "AVG(efficiency)": 0.011304347826087,
                  "subject": "2017",
                  "fullMark": 150,
                },
                "type": "circle",
                "value": "2017",
              },
            ]
          }
          verticalAlign="middle"
          wrapperStyle={
            Object {
              "maxHeight": 400,
              "width": 0,
            }
          }
        />,
        false,
      ],
      "height": 400,
      "width": 800,
    },
    Object {},
  ],
  "RadarProps": undefined,
}
`;

exports[`plugin-chart-piw Chart with legend top / percentage labels  1`] = `
Object {
  "LegendProps": Array [
    Object {
      "align": "center",
      "iconSize": 10,
      "iconType": "square",
      "layout": "horizontal",
      "onClick": [Function],
      "payload": Array [
        Object {
          "color": "#1f77b4",
          "id": "2020",
          "payload": Object {
            "AVG(efficiency)": 0.0566666666666667,
            "subject": "2020",
            "fullMark": 150,
          },
          "type": "circle",
          "value": "2020",
        },
        Object {
          "color": "#ff7f0e",
          "id": "2018",
          "payload": Object {
            "AVG(efficiency)": 0.038,
            "subject": "2018",
            "fullMark": 150,
          },
          "type": "circle",
          "value": "2018",
        },
        Object {
          "color": "#2ca02c",
          "id": "2019",
          "payload": Object {
            "AVG(efficiency)": 0.0162068965517241,
            "subject": "2019",
            "fullMark": 150,
          },
          "type": "circle",
          "value": "2019",
        },
        Object {
          "color": "#d62728",
          "id": "Total subject",
          "payload": Object {
            "AVG(efficiency)": 0.0138888888888889,
            "subject": "Total subject",
            "fullMark": 150,
          },
          "type": "circle",
          "value": "Total subject",
        },
        Object {
          "color": "#9467bd",
          "id": "2017",
          "payload": Object {
            "AVG(efficiency)": 0.011304347826087,
            "subject": "2017",
            "fullMark": 150,
          },
          "type": "circle",
          "value": "2017",
        },
      ],
      "verticalAlign": "top",
      "wrapperStyle": Object {
        "maxHeight": 400,
      },
    },
    Object {},
  ],
  "RadarChartProps": Array [
    Object {
      "children": Array [
        <mockConstructor
          align="center"
          iconSize={10}
          iconType="square"
          layout="horizontal"
          onClick={[Function]}
          payload={
            Array [
              Object {
                "color": "#1f77b4",
                "id": "2020",
                "payload": Object {
                  "AVG(efficiency)": 0.0566666666666667,
                  "subject": "2020",
                  "fullMark": 150,
                },
                "type": "circle",
                "value": "2020",
              },
              Object {
                "color": "#ff7f0e",
                "id": "2018",
                "payload": Object {
                  "AVG(efficiency)": 0.038,
                  "subject": "2018",
                  "fullMark": 150,
                },
                "type": "circle",
                "value": "2018",
              },
              Object {
                "color": "#2ca02c",
                "id": "2019",
                "payload": Object {
                  "AVG(efficiency)": 0.0162068965517241,
                  "subject": "2019",
                  "fullMark": 150,
                },
                "type": "circle",
                "value": "2019",
              },
              Object {
                "color": "#d62728",
                "id": "Total subject",
                "payload": Object {
                  "AVG(efficiency)": 0.0138888888888889,
                  "subject": "Total subject",
                  "fullMark": 150,
                },
                "type": "circle",
                "value": "Total subject",
              },
              Object {
                "color": "#9467bd",
                "id": "2017",
                "payload": Object {
                  "AVG(efficiency)": 0.011304347826087,
                  "subject": "2017",
                  "fullMark": 150,
                },
                "type": "circle",
                "value": "2017",
              },
            ]
          }
          verticalAlign="top"
          wrapperStyle={
            Object {
              "maxHeight": 400,
            }
          }
        />,
        <mockConstructor
          activeIndex={0}
          cx="50%"
          data={
            Array [
              Object {
                "AVG(efficiency)": 0.0566666666666667,
                "subject": "2020",
                "fullMark": 150,
              },
              Object {
                "AVG(efficiency)": 0.038,
                "subject": "2018",
                "fullMark": 150,
              },
              Object {
                "AVG(efficiency)": 0.0162068965517241,
                "subject": "2019",
                "fullMark": 150,
              },
              Object {
                "AVG(efficiency)": 0.0138888888888889,
                "subject": "Total subject",
                "fullMark": 150,
              },
              Object {
                "AVG(efficiency)": 0.011304347826087,
                "subject": "2017",
                "fullMark": 150,
              },
            ]
          }
          dataKey="AVG(efficiency)"
          label={[Function]}
          onClick={[Function]}
          outerRadius={100}
        >
          <Cell
            fill="#1f77b4"
          />
          <Cell
            fill="#ff7f0e"
          />
          <Cell
            fill="#2ca02c"
          />
          <Cell
            fill="#d62728"
          />
          <Cell
            fill="#9467bd"
          />
        </mockConstructor>,
      ],
      "height": 400,
      "width": 800,
    },
    Object {},
  ],
  "RadarProps": Array [
    Object {
      "activeIndex": 0,
      "children": Array [
        <Cell
          fill="#1f77b4"
        />,
        <Cell
          fill="#ff7f0e"
        />,
        <Cell
          fill="#2ca02c"
        />,
        <Cell
          fill="#d62728"
        />,
        <Cell
          fill="#9467bd"
        />,
      ],
      "cx": "50%",
      "data": Array [
        Object {
          "AVG(efficiency)": 0.0566666666666667,
          "subject": "2020",
          "fullMark": 150,
        },
        Object {
          "AVG(efficiency)": 0.038,
          "subject": "2018",
          "fullMark": 150,
        },
        Object {
          "AVG(efficiency)": 0.0162068965517241,
          "subject": "2019",
          "fullMark": 150,
        },
        Object {
          "AVG(efficiency)": 0.0138888888888889,
          "subject": "Total subject",
          "fullMark": 150,
        },
        Object {
          "AVG(efficiency)": 0.011304347826087,
          "subject": "2017",
          "fullMark": 150,
        },
      ],
      "dataKey": "AVG(efficiency)",
      "label": [Function],
      "onClick": [Function],
      "outerRadius": 100,
    },
    Object {},
  ],
}
`;
